### ISSUE TITLE
How to subset samples by their ids?

### ISSUE THREAD

**Issue by garyzhubc** [privefl/bigsnpr#301]
I have a bigSNP object and want to do some matrix operations by taking only a subset of samples in a list of ids that I am storing in a character vector. How am I supposed to do this? (maybe creating a new object that's only these ids?)

**Comment by privefl** [comment]
You'll need to use `match()` to get the corresponding row indices.
Then you can use as parameter `ind.row` in the {bigstatsr} functions that perform matrix multiplication.

**Comment by garyzhubc** [comment]
`mathc()` requires one list of ids that I have already, are you saying I can use that on the bigSNP object? 

I think I need to do a subtraction operation on the genotype matrix, namely to do `X[test_ids,:]-sum(X[control_ids,:], axis=0)` in Python

**Comment by privefl** [comment]
Yes, use also what you have in `$fam`.

If you use the right function for scaling in the PCA functions, then the center will be done for you, you should not have to transform the genotype matrix yourself (it should be the case for most functions of the packages).

**Comment by privefl** [comment]
There are two tutorials on computing PCA at https://privefl.github.io/bigsnpr/; have you looked at them?

**Comment by garyzhubc** [comment]
bigSNP object doesn't seem to have `$fam` filed and other fields doesn't seem to have sample ID. Is it because I'm using bgen instead of bed? I'm not doing exactly PCA because I need to separate case and control. 
![image](https://user-images.githubusercontent.com/30366447/150698115-bcb461b5-2d5a-4c6f-8477-37aff48f2143.png)

This is how I loaded the data
```
info <- snp_readBGI("chr_maf_001.bgen.bgi")
snp_id <- with(info, paste(chromosome, position, allele1, allele2, sep = "_"))
chr <- snp_readBGEN("chr_maf_001.bgen","chr",list(snp_id))
obj.bigSNP <- snp_attach(chr)
```

**Comment by privefl** [comment]
You need to read this information from the associated `.sample` file then.
Beware that the first line is the header (column names) and the second line is the types.
I typically use `bigreadr::fread2( )[-1, ]` to read it.

**Comment by garyzhubc** [comment]
Do you mean to read the sample file into another vector and integrate that into the bigSNP object? If so, how does the integration happen, given that I have a dataframe `sample <- bigreadr::fread2("chr_maf_001.sample")[-1,]`? Should I just use `obj.bigSNP$sample.ID <- sample$ID_2`
![image](https://user-images.githubusercontent.com/30366447/150701850-96425d4b-dd05-445e-baf3-bec0d9702d82.png)

**Comment by privefl** [comment]
You don't have to integrate it per se.
But you can read the IDs and use that in `match()`.

**Comment by garyzhubc** [comment]
I see, so using `match()` to get the number index from the character index. So I did
```
sample <- bigreadr::fread2("chr_maf_001.sample")[-1,]
ind.row <- match(paste0(control_id,"_1"), sample$ID_2)
```
Should I do `apply` on genotype matrix like 
```
apply(obj.bigSNP$genotypes[ind.row,], 2, mean)
```
because it seems quite slow and not sure if it's gonna finish in a reasonable time. Is there any efficient routines in the package that I can call?

**Comment by garyzhubc** [comment]
Nevermind I think I found `big_apply` seems will do the job

**Comment by privefl** [comment]
There are other functions as well.
Have a look at the websites of the packages, the lists of functions there and the tutorials, to get a sense of what you can use.

**Comment by garyzhubc** [comment]
It seems that if I do 

```
control_means <- big_apply(obj.bigSNP$genotypes, function(X, ind) apply(X[ind, ], 2, mean), ind=control_id_row)
case_matrix <- big_apply(obj.bigSNP$genotypes, function(X, ind) X[ind, ]-control_means, ind=rows_along(obj.bigSNP$genotypes))
```

this will create a large `case_matrix` in the R environment, which I think should be avoided ideally. Is there a way I can modify the bigSNP object itself, instead of creating a new matrix?

**Comment by privefl** [comment]
Why do you want to make a centered version of the genotype matrix? 
There is usually no need to make that explicitly.

**Comment by garyzhubc** [comment]
Yea I'm modifying the PCA for a case-control study, so the effect sizes (data) are properly weighted by case-control difference (deviation of case to average control), which is slightly different from the ancestral PCA. But I guess all methods here can only export R vectors, instead of modifying the genotype matrix?

**Comment by garyzhubc** [comment]
I think the above doesn't really work.
```
> case_matrix <- as_FBM(big_apply(obj.bigSNP$genotypes, function(X, ind) sweep(X[ind, ], 2, control_means), ind=case_id_row), backingfile = "case_matrix")
Error in { : 
  task 1 failed - "task 1 failed - "non-numeric argument to binary operator""
```

**Comment by garyzhubc** [comment]
I think I got the error: 
```
> case_matrix <- as_FBM(big_apply(obj.bigSNP$genotypes, function(X, ind) sweep(X[ind, ], 2, control_means), ind=rows_along(obj.bigSNP$genotypes)), backingfile = "case_matrix")
Error in { : 
  task 1 failed - "task 1 failed - "non-numeric argument to binary operator""
```

**Comment by privefl** [comment]
Just use the `fun.scaling` parameter in the PCA functions.

E.g.
```r
library(bigsnpr)

obj.bigsnp <- snp_attachExtdata()
G <- obj.bigsnp$genotypes
y <- obj.bigsnp$fam$affection - 1
gwas <- big_univLogReg(G, y)

fun <- function(X, ind.row, ind.col) {
  data.frame(center = big_scale()(X, ind.row, ind.col)$center,
             scale = 1 / gwas$estim)
}
obj.svd <- big_randomSVD(G, fun.scaling = fun)

plot(obj.svd, type = "scores") + 
  ggplot2::aes(color = as.factor(y))
```

**Comment by privefl** [comment]
Maybe it would be easier if I would allow for providing directly a data.frame with center and scale, instead of having to tweak the function like this.

**Comment by garyzhubc** [comment]
Yes that would be helpful! I have the `control_means` vector already, so I really just need to subtract it off, and then `big_SVD` already allows me to normalize with std only on the cases. 

Is this feature gonna take a while to be added?

**Comment by privefl** [comment]
> big_SVD already allows me to normalize with std only on the cases.

What do you mean exactly?

**Comment by garyzhubc** [comment]
I mean I just need to scale-standardize the matrix to variance 1 with before PCA (with a modified center-standardization given in the vector above)

**Comment by garyzhubc** [comment]
Is it gonna be a big feature to add or is it just gonna be a small change? If there's anything I can help let me know.

**Comment by garyzhubc** [comment]
I just notice how this is done in scikit-learn of Python, maybe allowing the mean input is reasonable https://scikit-learn.org/stable/modules/generated/sklearn.preprocessing.StandardScaler.html#sklearn.preprocessing.StandardScaler

**Comment by garyzhubc** [comment]
> I think I got the error using the procedure above (broadcasting matrix-vector subtraction using `sweep`):
> 
> ```
> > case_matrix <- as_FBM(big_apply(obj.bigSNP$genotypes, function(X, ind) sweep(X[ind, ], 2, control_means), ind=rows_along(obj.bigSNP$genotypes)), backingfile = "case_matrix")
> Error in { : 
>   task 1 failed - "task 1 failed - "non-numeric argument to binary operator""
> ```

I also still wonder why this is not working... I might happen to have enough memory to store the matrix if it's below 250G, although this isn't gonna be the ideal way to do this that leverages the power of this package. The bgen file has only 1G, so this is promising, too.

**Comment by privefl** [comment]
Okay, I think I got this working:
```r
as_scaling_fun <- function(center.col, scale.col, ind.col = seq_along(center.col)) {
  
  bigassertr::assert_lengths(center.col, scale.col, ind.col)
  
  e <- new.env(parent = baseenv())
  assign("_DF_", data.frame(center = center.col, scale = scale.col), envir = e)
  assign("_IND_COL_", ind.col, envir = e)
  
  f <- function(X, ind.row, ind.col) {
    ind <- match(ind.col, `_IND_COL_`)
    `_DF_`[ind, ]
  }
  environment(f) <- e
  
  f
}
```

With this, you can create the appropriate `fun.scaling` parameter easily.

I'll add this function to {bigstatsr} for convenience when I have more time later.

**Comment by privefl** [comment]
In your `big_apply()`, you are missing `a.combine = "rbind"`. 

Also, try to operate on a subset of columns, instead of a subset of rows, that's much more efficient.

And try to never make these new large objects, that's completely against the mindset of my packages.

**Comment by garyzhubc** [comment]
> Okay, I think I got this working:
> 
> ```r
> as_scaling_fun <- function(center.col, scale.col, ind.col = seq_along(center.col)) {
>   
>   bigassertr::assert_lengths(center.col, scale.col, ind.col)
>   
>   e <- new.env(parent = baseenv())
>   assign("_DF_", data.frame(center = center.col, scale = scale.col), envir = e)
>   assign("_IND_COL_", ind.col, envir = e)
>   
>   f <- function(X, ind.row, ind.col) {
>     ind <- match(ind.col, `_IND_COL_`)
>     `_DF_`[ind, ]
>   }
>   environment(f) <- e
>   
>   f
> }
> ```
> 
> With this, you can create the appropriate `fun.scaling` parameter easily.
> 
> I'll add this function to {bigstatsr} for convenience when I have more time later.
I tried doing this 
`obj.svd <- big_SVD(obj.bigSNP$genotypes, fun.scaling = as_scaling_fun, control_means, case_std)`

but got error `Error: 'fun.scaling' should have arguments named 'ind.row, ind.col'.`

**Comment by privefl** [comment]
You should do something like:
```r
G <- obj.bigSNP$genotypes
control_means <- big_scale()(G, control_id_row)$center
case_std <- big_scale()(G, case_id_row)$scale
obj.svd <- big_randomSVD(G, fun.scaling = as_scaling_fun(control_means, case_std))
```
